File: ios/PlatformChannelPerformance/infinitescroll/components/GetItemsUseCaseFlutterWrapper.swift
//
//  GetItemsUseCaseFlutterWrapper.swift
//  PlatformChannelPerformance
//
//  Created by Markus Reinhold on 09.11.23.
//

import Foundation
import Combine

class GetItemsUseCaseFlutterWrapper: NSObject, FlutterStreamHandler {
    private var getItemsUseCase: GetItemsUseCase?
    
    private var subscription: AnyCancellable?
    
    init(flutterBinaryMessenger: FlutterBinaryMessenger) {
        super.init()
        FlutterEventChannel(name: "scrollItemStream", binaryMessenger: flutterBinaryMessenger).setStreamHandler(self)
        FlutterMethodChannel(name: "scrollItemRequest", binaryMessenger: flutterBinaryMessenger).setMethodCallHandler(self.onMethodCall)
    }
    
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        getItemsUseCase = GetItemsUseCase()
        getItemsUseCase?.loadContent(targetPage: PAGE_PRELOAD_COUNT)
        
        subscription = getItemsUseCase?.$items
            .sink(receiveValue: { items in
                events(items.toFlutter())
            })
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        subscription?.cancel()
        getItemsUseCase = nil
        return nil
    }
    
    func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
        switch call.method {
        case "loadContent":
            if let targetPage = call.arguments as? Int {
                getItemsUseCase?.loadContent(targetPage: targetPage)
                result(nil)
            } else {
                result(FlutterError(code: "bad args", message: nil, details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}


extension Array where Element == ScrollItem {
    func toFlutter() -> [Any] {
        return map { $0.toFlutter() }
    }
}

extension ScrollItem {
    func toFlutter() -> [Any] {
        var flutterList = [Any]()
        flutterList.append(id)
        flutterList.append(header)
        return flutterList
    }
}

File: android/benchmarkable/src/main/java/com/example/platformchannelperformance/infinitescroll/GetItemsUseCaseFlutterWrapper.kt
package com.example.platformchannelperformance.infinitescroll

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class GetItemsUseCaseFlutterWrapper(binaryMessenger: BinaryMessenger) :
    EventChannel.StreamHandler,
    MethodChannel.MethodCallHandler {

    private var getItemsUseCase: GetItemsUseCase? = null
    private var coroutineScope: CoroutineScope? = null

    init {
        EventChannel(
            binaryMessenger,
            "scrollItemStream"
        ).setStreamHandler(this@GetItemsUseCaseFlutterWrapper)
        MethodChannel(
            binaryMessenger,
            "scrollItemRequest"
        ).setMethodCallHandler(this@GetItemsUseCaseFlutterWrapper)
    }

    // EventChannel
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        getItemsUseCase = GetItemsUseCase().also { useCase ->
            useCase.loadNewItems(PAGE_PRELOAD_COUNT)
            coroutineScope?.launch {
                useCase.items.collect {
                    withContext(Dispatchers.Main) {
                        events?.success(it.toFlutter())
                    }
                }
            }
        }
    }

    // EventChannel
    override fun onCancel(arguments: Any?) {
        coroutineScope?.cancel()
        coroutineScope = null
        getItemsUseCase = null
    }

    // MethodChannel
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadContent" -> when (val targetPage = call.arguments as? Int) {
                null -> result.error("bad args", null, null)
                else -> {
                    getItemsUseCase?.loadNewItems(targetPage)?.also {
                        result.success(null)
                    } ?: result.error("not instantiated", null, null)
                }
            }

            else -> result.notImplemented()
        }
    }

    // converter functions

    private fun List<Item>.toFlutter() = map { it.toFlutter() }

    private fun Item.toFlutter() = buildList<Any> {
        add(id)
        add(header)
    }
}



